SAMLPortal/Controllers/AppController.cs

```diff
@@ -0,0 +1,219 @@
+using SAMLPortal.Misc;
+
+namespace SAMLPortal.Controllers
+{
+	[Route("App")]
+	[Authorize(Roles = UserRoles.Administrator)]
+	public class AppController : Controller
+	{
+		// GET: App
+		public ActionResult Index()
+		{
+			SAMLPortalContext context = new SAMLPortalContext();
+			return View(context.App.ToList());
+		}
+
+		// GET: App/Details/5
+		[Route("Details/{id}")]
+		public ActionResult Details(int id)
+		{
+			SAMLPortalContext context = new SAMLPortalContext();
+			App app = context.App.Find(id);
+			return View(app);
+		}
+
+		// GET: App/Create
+		[HttpGet]
+		[Route("Create")]
+		public ActionResult Create()
+		{
+			App emptyApp = new App();
+			return View(emptyApp);
+		}
+
+		// Post: App/Create
+		[HttpPost]
+		[ValidateAntiForgeryToken]
+		[Route("Create")]
+		public async Task<IActionResult> Create(App appli)
+		{
+			try
+			{
+				SAMLPortalContext context = new SAMLPortalContext();
+				if (ModelState.IsValid)
+				{
+					if (appli.MetadataURL != null)
+					{
+
+					}
+					context.Add(appli);
+					await context.SaveChangesAsync();
+					return Redirect("/");
+				}
+			}
+			catch (DbUpdateException /* ex */)
+			{
+				//Log the error (uncomment ex variable name and write a log.
+				ModelState.AddModelError("", "Unable to save changes. " +
+					"Try again, and if the problem persists " +
+					"see your system administrator.");
+			}
+			return View(appli);
+		}
+
+		// GET: App/Edit/5
+		[AllowAnonymous]
+		[Route("Edit/{id}")]
+		public async Task<IActionResult> Edit(int? id)
+		{
+			if (id == null)
+			{
+				return NotFound();
+			}
+
+			SAMLPortalContext context = new SAMLPortalContext();
+			var app = await context.App.FindAsync(id);
+			if (app == null)
+			{
+				return NotFound();
+			}
+			return View("Edit", app);
+		}
+
+		// POST: App/Edit/5
+		[HttpPost]
+		[ValidateAntiForgeryToken]
+		[Route("Edit/{id}")]
+		public async Task<IActionResult> EditPost(int? id)
+		{
+			if (id == null)
+			{
+				return NotFound();
+			}
+			SAMLPortalContext context = new SAMLPortalContext();
+			var appToUpdate = await context.App.FirstOrDefaultAsync(s => s.Id == id);
+			if (await TryUpdateModelAsync<App>(
+				appToUpdate,
+				"",
+				s => s.Name,
+				s => s.Description,
+				s => s.Role,
+				s => s.Enabled,
+				s => s.MetadataURL,
+				s => s.Issuer,
+				s => s.SingleSignOnDestination,
+				s => s.SingleLogoutResponseDestination))
+			{
+				try
+				{
+					await context.SaveChangesAsync();
+					return Redirect("/");
+				}
+				catch (DbUpdateException /* ex */)
+				{
+					//Log the error (uncomment ex variable name and write a log.)
+					ModelState.AddModelError("", "Unable to save changes. " +
+						"Try again, and if the problem persists, " +
+						"see your system administrator.");
+				}
+			}
+			return View("Edit", appToUpdate);
+		}
+
+		// GET: App/Delete/5
+		[Route("Delete/{id}")]
+		public async Task<IActionResult> Delete(int? id, bool? saveChangesError = false)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use the overloading mechanism instead of the optional parameters.](https://app.codacy.com/manual/zegorax/SAMLPortal/pullRequest?prid=5434886)